src/main/java/de/tum/cit/aet/artemis/programming/service/localci/LocalCIEventListenerService.java

```diff
@@ -0,0 +1,182 @@
+import org.springframework.stereotype.Service;
+
+import com.hazelcast.collection.IQueue;
+import com.hazelcast.collection.ItemEvent;
+import com.hazelcast.collection.ItemListener;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.map.IMap;
+import com.hazelcast.map.listener.EntryAddedListener;
+import com.hazelcast.map.listener.EntryRemovedListener;
+import com.hazelcast.map.listener.EntryUpdatedListener;
+
+import de.tum.cit.aet.artemis.buildagent.dto.BuildAgentInformation;
+import de.tum.cit.aet.artemis.buildagent.dto.BuildJobQueueItem;
+import de.tum.cit.aet.artemis.programming.domain.build.BuildJob;
+import de.tum.cit.aet.artemis.programming.domain.build.BuildStatus;
+import de.tum.cit.aet.artemis.programming.dto.SubmissionProcessingDTO;
+import de.tum.cit.aet.artemis.programming.repository.BuildJobRepository;
+import de.tum.cit.aet.artemis.programming.service.ProgrammingMessagingService;
+
+@Service
+@Profile("localci & scheduling")
+public class LocalCIEventListenerService {
+
+    private static final Logger log = LoggerFactory.getLogger(LocalCIEventListenerService.class);
+
+    private final HazelcastInstance hazelcastInstance;
+
+    private final LocalCIQueueWebsocketService localCIQueueWebsocketService;
+
+    private final BuildJobRepository buildJobRepository;
+
+    private final SharedQueueManagementService sharedQueueManagementService;
+
+    private final ProgrammingMessagingService programmingMessagingService;
+
+    public LocalCIEventListenerService(@Qualifier("hazelcastInstance") HazelcastInstance hazelcastInstance, LocalCIQueueWebsocketService localCIQueueWebsocketService,
+            BuildJobRepository buildJobRepository, SharedQueueManagementService sharedQueueManagementService, ProgrammingMessagingService programmingMessagingService) {
+        this.hazelcastInstance = hazelcastInstance;
+        this.localCIQueueWebsocketService = localCIQueueWebsocketService;
+        this.buildJobRepository = buildJobRepository;
+        this.sharedQueueManagementService = sharedQueueManagementService;
+        this.programmingMessagingService = programmingMessagingService;
+    }
+
+    /**
+     * Add listeners for build job, build agent changes.
+     */
+    @PostConstruct
+    public void init() {
+        IQueue<BuildJobQueueItem> queue = hazelcastInstance.getQueue("buildJobQueue");
+        IMap<Long, BuildJobQueueItem> processingJobs = hazelcastInstance.getMap("processingJobs");
+        IMap<String, BuildAgentInformation> buildAgentInformation = hazelcastInstance.getMap("buildAgentInformation");
+        queue.addItemListener(new QueuedBuildJobItemListener(), true);
+        processingJobs.addEntryListener(new ProcessingBuildJobItemListener(), true);
+        buildAgentInformation.addEntryListener(new BuildAgentListener(), true);
+    }
+
+    /**
+     * Check the status of pending build jobs. If a build job is missing from the queue, not being built or not finished, update the status to missing.
```

Comment:
please improve the JavaDoc and explain why this scheduled task is actually necessary. Please be as extensive as possible so other developers understand the rationale behind it. Also relate it to boundary use cases, like restarting Artemis while there are still builds in the queue